### ISSUE TITLE
Changing default METHOD

### ISSUE THREAD

**Issue by permcody** [idaholab/moose#7341]
### Description of the enhancement or error report

I think it's time to talk about changing the default METHOD that people use with MOOSE. The current default of "opt" really isn't useful for developing and a lot of hair pulling and common errors would go away if people used the right mode for the kind of work they are performing.

### Rationale for the enhancement or information for reproducing the error

No error - just a change in practice

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

Everyone

**Comment by dschwen** [comment]
To what? `oprof`?

**Comment by friedmud** [comment]
`oprof` or `dbg` executables/libraries are huge and take forever to build on shared filesystems.  They also take up WAY more memory.

Also: I don't want a slew of "I just installed MOOSE for the first time and it's being really slow to solve a 2D diffusion problem" emails.

What do you gain with `oprof` or `dbg` executables for the "normal" errors that would stop "hair pulling out"?  For the record I actually do ALL of my development with `opt` executables because they are so much faster to build, leaner on memory and faster to run.  I only build `dbg` when I need to run a debugger...

Very against this change.

**Comment by YaqiWang** [comment]
I actually like the idea because as exactly @friedmud said I want to avoid ' build dbg when I need to run a debugger'. Whenever I decided to do that, I was reminded to stand up and go get a cup of coffee. It could be better to make developer able to set the default METHOD when building libmesh. Then users are not impacted. Is this possible?

**Comment by permcody** [comment]
Don't lump oprof and dbg together. They are drastically different in size and speed. 

oprof keeps symbols so you immediately have stack trace information, even in the hands on training people are immediately stuck when they write bad code, they don't get anything useful post mortem.  oprof is slightly bigger but basically just as fast once it's loaded. We have to understand our users and moving to oprof make a lot of sense for the vast majority. I'm willing to put some speed numbers up as I'm fairly certain your claims of it being big and slow (even on shared filesystems) aren't really well backed up for all but the biggest of runs on the crappiest of systems.

I found out something pretty cool on OS X. Whenever you get a seg fault, the results of the backtrace of the crash are automatically saved in the console. You don't even have to fire up the debugger to see why! It's always there. Doesn't work with opt mode.

At the very least we need to train people and get them using some of these other modes. Development in `opt` is actually pretty lame when you have such fast and powerful hardware at your fingertips.

**Comment by friedmud** [comment]
Building static executables now to see the size difference...

I'll run some compiling timings as well...

**Comment by permcody** [comment]
Static is not realistic, basically nobody uses that anymore. Yeah there's some guy doing a production run on a big cluster that is, but there's no need to compare that case since it's not used anywhere in practice

**Comment by friedmud** [comment]
Static shows the true size.  The same size as the in memory image.  Hard to get a handle on that with shared.

**Comment by friedmud** [comment]
The reason you think the binary sizes aren't much different is because you're looking at the _shared_ executables...

**Comment by permcody** [comment]
I'm actually going to challenge that as well. I don't believe that static shows true size. Let's take the "modules" executable for instance. It contains at least a dozen different libraries from moose to the modules. If I run an example in phase-field with combined module, presumably none of the other dynamic libraries would need to be loaded since there are no objects used from those libraries.

**Comment by friedmud** [comment]
Ok - once I get done building static (because it's already going) I'll run some numbers on the in-memory size of various MOOSE-based executables with shared.

**Comment by friedmud** [comment]
But - regardless... the dynamic library total _storage_ size must be the same (or similar) to the static executable size.

**Comment by permcody** [comment]
Let's at least recognize that we have users and developers. A one size fits all approach doesn't really work well. I certainly don't want to make things harder for users, but I don't want to keep impending or creating so much extra work for developers (who make up a large part of our community)

**Comment by friedmud** [comment]
I can agree with that - but the answer is documentation and training - not changing the default to unoptimized binaries.  Memory usage has always been one of the largest issues our users face... especially on clusters but also on workstations.

We could make `update_and_rebuild_libmesh.sh` interactive... have it ask which types of executables the user wants to create and explain the differences between them.  We should also provide better documentation of the various modes and the reasons to choose them (on the Wiki or in the new documentation system).

**Comment by permcody** [comment]
@jwpeterson will get right on that - he's the new `update_and_rebuild_libmesh.sh` king.

oprof isn't unoptimized 😄

**Comment by friedmud** [comment]
> oprof isn't unoptimized

Memory/storage-wise it is.

I switched away from oprof for three reasons:

1.  Longer build on HPC
2.  MUCH longer build on my distributed build system
3.  Memory use on HPC

**Comment by friedmud** [comment]
For me, `oprof` is the worst of both worlds.  It's bigger/slower than `opt`  I still need to build `dbg` when I'm actually running a debugger because I don't want optimization reordering stuff and I want all the extra checking of a debugger.

**Comment by permcody** [comment]
I actually switched away from `opt` too. However in my case I'm working on `devel` for day to day work fore several reasons

1. Icecream distributed build is fast (faster for opt but still under the pain threshold) 
2. When doing tricky framework coding, assertions are really, really handy (basically I don't see seg faults)
3. I don't do a lot of production runs on my workstation, when I do I can build opt really quickly.

**Comment by friedmud** [comment]
Static `oprof` and `dbg` are so huge that they can't be built on my laptop.  `ar` goes over 4GB of RAM and stays there until it spits out this:

```
fatal error: /Library/Developer/CommandLineTools/usr/bin/ar: fatal error in /Library/Developer/CommandLineTools/usr/bin/ranlib
```

I guess there must be some limit on executable size on OSX.

The libraries do get created though.  Here are the framework libraries:

```
-rw-r--r--  1 gastdr  staff   4.2G Jul 14 18:47 libmoose-dbg.a
-rw-r--r--  1 gastdr  staff   4.0G Jul 14 18:37 libmoose-oprof.a
-rw-r--r--  1 gastdr  staff    38M Jul 14 18:30 libmoose-opt.a
```

And for `moose_test`

```
-rw-r--r--  1 gastdr  staff   1.5G Jul 14 18:48 libmoose_test-dbg.a
-rw-r--r--  1 gastdr  staff   1.4G Jul 14 18:37 libmoose_test-oprof.a
-rw-r--r--  1 gastdr  staff    11M Jul 14 18:30 libmoose_test-opt.a
```

It's obvious that `dbg`/`oprof` use at least two orders of magnitude more storage (also, note that oprof uses about the same as `dbg`).

Now, let me look at the in-memory size of a dynamically linked executable... BRB.

**Comment by jwpeterson** [comment]
These numbers are pretty interesting!  I figured oprof would be bigger than opt, but I'm surprised it's *that* much bigger.

**Comment by permcody** [comment]
These are all static. When compiling in dynamic mode all of the libraries are much closer in size. Yes I know that static mode produces ridiculously sized binaries. Now if there's a significant different in the in memory size...

**Comment by permcody** [comment]
Isn't this all a case of "premature optimization"? Why do we really care that these binaries are a little bigger and slower? What if they save people time rather than cost them time? It all depends on what people are doing. 

At least I know Derek's opinion.

**Comment by friedmud** [comment]
> little bigger

I don't think that means what you think it means.  It certainly doesn't mean two orders of magnitude...

Dynamically linked results coming up soon...

**Comment by friedmud** [comment]
Ok - I believe you.  Go ahead and switch to `oprof` by default if you like.

In-memory image for the executable is not much different when doing dynamic linking.  Usually a difference of 5-10%.  I cannot at all account for why the difference is so small when the static image of the executables is orders of magnitude different... but I also can't deny what I see.  My only guess is that the OS is WAY better at partial loading of dynamic libraries than I assumed.  I figured that if a library was touched at all then the whole thing would be loaded... it appears that's not the case (which is really cool).

For the record, I tried moose_test, rattlesnake and bison: all with the same result.

The only difference that I can see is that oprof _does_ take longer for me to compile with my compile farm here.  For dynamically linked compile timing for moose/moose_test (using my distributed build system here from my laptop):

```
opt: 194.74
oprof: 297.32
dbg: 317.13
```

So.  Yeah.  You win.  I was wrong.  Move forward ;-)

**Comment by permcody** [comment]
The compile times for oprof are a little worse than I expected. I think
we'll run some tests on single machine compiles to see if they are closer.
People do complain about compile times so I don't want to add another 30%
without thinking about it

**Comment by friedmud** [comment]
The distributed build system I'm using should exacerbate the difference. I
doubt you'll find 30% difference in single machine compiling.

That said, we should run some compile time comparisons on Falcon as well.
The mounted filesystems may have more trouble keeping up with the larger .o
files generated by oprof.

**Comment by lindsayad** [comment]
@permcody Have you put up any documentation up on the wiki about the different modes? I assume that if I want to use my debugger and step through my code, printing variables here and there, that I still need to compile in `dbg` as opposed to `devel`?

**Comment by permcody** [comment]
I have not...  I will eventually :)

It depends on what level you need in debugging. "devel" still passes -O3 which creates an optimized binary but with lots of additional options such as assertions and all symbols. It's nice because it runs reasonably fast but still gives you line numbers and things like bounds checks (through the use of our plentiful assertions in MOOSE and libMesh) as you develop. If you really need to go line by line though, it may not be the mode for you due to the optimized flag option.

**Comment by dschwen** [comment]
Alex, the libmesh README has a list: https://github.com/libMesh/libmesh
